### ISSUE TITLE
JumpTable::PatchCallSites() dangerous casts

### ISSUE THREAD

**Issue by danluu** [BitFunnel/NativeJIT#54]
~~~
    // WARNING: Non portable. Assumes little endian machine architecture.                                                                                                                
    // WARNING: Non portable. Assumes that fixup value is labelAddress - siteAddress - size.Size().                                                                                      
    void JumpTable::PatchCallSites()
    {                                                                                                                                                                            
        for (size_t i=0; i < m_callSites.size(); ++i)
        {                                                                                                                                                                        
            const CallSite& site = m_callSites[i];
            const uint8_t* labelAddress = AddressOfLabel(site.GetLabel());
            uint8_t* siteAddress = site.Site();
            size_t delta = labelAddress - siteAddress - site.Size();

            // TODO: Evaluate whether special cases for size == 2 and size == 4 actually improve performance.                                                                            
            size_t size = site.Size();
            if (size == 2)
            {
                *((int16_t*)siteAddress) = (int16_t)delta;
                siteAddress += size;
            }
            else if (size == 4)
            {
                *((int32_t*)siteAddress) = (int32_t)delta;
                siteAddress += size;
            }
            else
            {
                while (size > 0)
                {
                    *siteAddress++ = (uint8_t)delta;
                    delta = delta >> 8;
                    size--;
                }
            }
        }
    }
~~~

I'm enabling `-Wold-style-cast` (see BitFunnel/BitFunnel#37) and converting these to C++ style casts, but this function still seems dangerous.

**Comment by danluu** [comment]
I believe the TODOs here are to replace `size_t` with `ptrdiff_t` and to at least check that casting to a smaller size won't overflow the size and `throw` in case of an overflow.

We could actually handle deltas that are too large, but given that this code has been used in production for a while without issue, it doesn't seem like we need that capability.

I'll submit a fix for this after the change that's out for code reviewing (which would merge conflict with this) gets approved.

**Comment by danluu** [comment]
Fixed in 83dc5bbfe0b2596035383950b5d0eb07986e819e now that the other change has gone in.
